framework/src/org/checkerframework/common/value/ValueAnnotatedTypeFactory.java

```diff
@@ -580,26 +636,79 @@ private void handleDimensions(
             if (AnnotationUtils.areSameIgnoringValues(dimType, BOTTOMVAL)) {
                 type.replaceAnnotation(BOTTOMVAL);
             } else {
-                List<Long> longLengths = null;
+                RangeOrListOfValues rolv = null;
                 if (AnnotationUtils.areSameByClass(dimType, IntRange.class)) {
-                    longLengths =
-                            ValueCheckerUtils.getValuesFromRange(getIntRange(dimType), Long.class);
+                    rolv = new RangeOrListOfValues(getRange(dimType));
                 } else if (AnnotationUtils.areSameByClass(dimType, IntVal.class)) {
-                    longLengths = getIntValues(dimType);
+                    rolv = new RangeOrListOfValues();
+                    rolv.addAll(rolv.convertLongsToInts(getIntValues(dimType)));
                 }
-                if (longLengths != null) {
-                    HashSet<Integer> lengths = new HashSet<Integer>(longLengths.size());
-                    for (Long l : longLengths) {
-                        lengths.add(l.intValue());
+                if (rolv != null) {
+                    AnnotationMirror newQual;
+                    if (rolv.isRange) {
+                        newQual = createArrayLenRangeAnnotation(rolv.range);
```

Comment:
Make createAnnotation a method in RangeOrListOfValues. There are two occurs that of this pattern that can be replaced with a call to that method.